File: frontend/src/types/policy.ts
import { PolicyStatus, UUID } from './common'

export type PolicyCategory = 
  | 'VELOCITY'
  | 'AMOUNT'
  | 'IDENTITY'
  | 'PAYMENT_INSTRUMENT'
  | 'REFUNDS'
  | 'PROMOTIONS'
  | 'BEHAVIORAL'

export type PolicyRuleType =
  | 'VELOCITY_ACCOUNT'
  | 'VELOCITY_DEVICE'
  | 'VELOCITY_INSTRUMENT'
  | 'VELOCITY_ADDRESS'
  | 'VELOCITY_IP'
  | 'AMOUNT_MAX'
  | 'AMOUNT_DAILY'
  | 'AMOUNT_WEEKLY'
  | 'IDENTITY_ACCOUNT_AGE'
  | 'IDENTITY_DEVICE_COUNT'
  | 'IDENTITY_IP_CHANGES'
  | 'IDENTITY_ADDRESS_REUSE'
  | 'INSTRUMENT_CARDS_PER_ACCOUNT'
  | 'INSTRUMENT_ACCOUNTS_PER_CARD'
  | 'INSTRUMENT_REUSE'
  | 'REFUND_FREQUENCY'
  | 'REFUND_RATIO'
  | 'REFUND_THRESHOLD'
  | 'PROMOTION_COUPON'
  | 'PROMOTION_REFERRAL'
  | 'PROMOTION_NEW_USER'
  | 'BEHAVIOR_RAPID_SWITCH'
  | 'BEHAVIOR_CHECKOUT_FAILURES'
  | 'BEHAVIOR_BURST'
  | 'BEHAVIOR_UNUSUAL_SEQUENCE'

export type RuleAction = 'BLOCK' | 'FLAG' | 'MONITOR'

export interface PolicyRule {
  id: UUID
  policyVersionId: UUID
  name: string
  ruleType: PolicyRuleType
  category: PolicyCategory
  parameters: Record<string, any>
  action: RuleAction
  isEnabled: boolean
  sequenceOrder: number
  description?: string
}

export interface PolicyVersion {
  id: UUID
  policyId: UUID
  versionNumber: string
  status: PolicyStatus
  rules: PolicyRule[]
  createdAt: string
  createdBy: string
  notes?: string
  effectivenessScore?: number
  attackCoverageScore?: number
}

export interface RiskPolicy {
  id: UUID
  merchantId: UUID
  name: string
  description: string
  category: PolicyCategory
  isActive: boolean
  currentVersionNumber: string
  currentVersionId: UUID
  ruleCount: number
  coverageRate: number // 0-100%
  effectivenessRate: number // 0-100%
  lastUpdated: string
  versions?: PolicyVersion[]
}

File: backend/app/schemas/policy.py
from enum import Enum
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field
from backend.app.schemas.common import PolicyStatus, SeverityLevel


class PolicyCategory(str, Enum):
    VELOCITY = "VELOCITY"
    AMOUNT = "AMOUNT"
    IDENTITY = "IDENTITY"
    PAYMENT_INSTRUMENT = "PAYMENT_INSTRUMENT"
    REFUNDS = "REFUNDS"
    PROMOTIONS = "PROMOTIONS"
    BEHAVIORAL = "BEHAVIORAL"


class PolicyRuleType(str, Enum):
    VELOCITY_ACCOUNT = "VELOCITY_ACCOUNT"
    VELOCITY_DEVICE = "VELOCITY_DEVICE"
    VELOCITY_INSTRUMENT = "VELOCITY_INSTRUMENT"
    VELOCITY_ADDRESS = "VELOCITY_ADDRESS"
    VELOCITY_IP = "VELOCITY_IP"
    AMOUNT_MAX = "AMOUNT_MAX"
    AMOUNT_DAILY = "AMOUNT_DAILY"
    IDENTITY_ACCOUNT_AGE = "IDENTITY_ACCOUNT_AGE"
    IDENTITY_DEVICE_COUNT = "IDENTITY_DEVICE_COUNT"
    INSTRUMENT_CARDS_PER_ACCOUNT = "INSTRUMENT_CARDS_PER_ACCOUNT"
    REFUND_RATIO = "REFUND_RATIO"
    PROMOTION_COUPON = "PROMOTION_COUPON"
    BEHAVIOR_RAPID_SWITCH = "BEHAVIOR_RAPID_SWITCH"


class RuleAction(str, Enum):
    BLOCK = "BLOCK"
    FLAG = "FLAG"
    MONITOR = "MONITOR"


class PolicyRuleSchema(BaseModel):
    id: Optional[str] = None
    policy_version_id: Optional[str] = None
    name: str
    rule_type: PolicyRuleType
    category: PolicyCategory
    parameters: Dict[str, Any] = Field(default_factory=dict)
    action: RuleAction
    is_enabled: bool = True
    sequence_order: int = 1
    description: Optional[str] = None


class PolicyVersionSchema(BaseModel):
    id: str
    policy_id: str
    version_number: str
    status: PolicyStatus
    rules: List[PolicyRuleSchema] = Field(default_factory=list)
    created_at: str
    created_by: str
    notes: Optional[str] = None


class PolicyCreate(BaseModel):
    name: str
    description: str
    category: PolicyCategory
    rules: List[PolicyRuleSchema]
    notes: Optional[str] = "Initial version"


class PolicyUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    is_active: Optional[bool] = None
    notes: Optional[str] = None


class PolicyResponse(BaseModel):
    id: str
    merchant_id: str
    name: str
    description: str
    category: PolicyCategory
    current_version_id: str
    current_version_number: str
    is_active: bool
    rule_count: int
    coverage_rate: float
    effectiveness_rate: float
    created_at: str
    updated_at: str
    versions: List[PolicyVersionSchema] = Field(default_factory=list)
